mitodl/ol-django
# Switch to using scriv for changelogs

Note: The line count on the PR appears intimidating, but lockfiles account for most of that. I added `.gitattributes` to map those as binary files going forward to eliminate that. So for this PR the diff count is actually:

```
❯ git diff main --shortstat
 78 files changed, 670 insertions(+), 337 deletions(-)
```

This PR closes #108. The scope creeped a bit on this one because it was a bit difficult to integrate it into our existing release scripting and I wanted to try to keep the ease-of-use at a similar level (e.g. runnable in a one-liner and not requiring the engineer to install libraries manually).

https://scriv.readthedocs.io/

- Switches our commands from `argparse` to [`click`](https://click.palletsprojects.com/)  and [`cloup`](https://cloup.readthedocs.io/) (`cloup` extends `click` with some utilties that are helpful)
  - Since `scriv` itself uses the `click` library, I opted to switch our own implementation over to that because it gave a few benefits:
    - `click` in general has better ergonomics
    - Our `click` commands can compose and invoke other `click` commands easily (otherwise we would've had to do some `multiprocessing` stuff here and parse outputs, which gets messy quick)
  - The code has additionally been split out into multiple commands and subcommands to make it easier to test or use specific ones in isolation.
- I decided to pull in the [`bumpver`](https://github.com/mbarkhau/bumpver) library and switch to date-based version numbers since this means we don't have to worry about deciding on a semantic versioning change. 
  - This tool replaces custom code we had to read and rewrite `__init__.py` - it offers a lot more flexibility for search/replace of version strings. It also supports automatically generated date-based versioning so we don't ever have to worry about input and this will be easier to automate. We'll be able to use this on our other projects too.
  - **NOTE:** I had to fake a "current" version number for all the packages that followed the date based format because `bumpver` was trying to parse them. I used the date of the last release.
  - I switched `first_party_dependency_version_scheme` in the pants config to `"any"` so that we don't have to constantly release libraries that don't actually have changes because the date changes (this problem we've hit before with semantic versioning, but date based versioning will make it worse.
- `pants` upgrade to a beta version (should be released soon) to pick up some bug fixes

To test this out, run the following commands:
- `./pants ol-project changelog list --app common` - this should inform you there are no changelog fragments.
- `./pants ol-project release create --app common --no-switch-main` - this should return the same message as the command above and not generate a release
- `./pants ol-project changelog create --app common`. - this will create a file using a template in `src/mitol/common/changelog.d/ that can be edited and committed. (You don't need to commit it for testing)
- `./pants ol-project changelog list --app common` - this list the fragments for the common app
- `./pants ol-project version show --app common` - this will display the current version.
- `./pants ol-project release create --app common --no-switch-main` - this will create a release. 
   - Normally we'd run it without `--no-switch-main`, but if you don't specify that it'll try to switch to that branch. 
   - We'd also run it with `--push`, but you don't want to do that here as it'll trigger an actual release to go out to pypi.
